src/rsa/signing.rs

```diff
@@ -340,22 +346,33 @@ impl RSAKeyPair {
 
                 // Step 7.b is done out-of-order below.
 
-                // Step 7.c.
-                let qInv = qInv.into_elem(&p.modulus)?;
-
-                // Steps 7.d and 7.e are omitted per the documentation above,
-                // and because we don't (in the long term) have a good way to
-                // do modulo with an even modulus.
-
-                // Step 7.f.
-                let qInv =
-                    bigint::elem_mul(p.oneRR.as_ref(), qInv, &p.modulus)?;
                 let q_mod_p = q.try_clone()?.into_elem(&p.modulus)?;
-                let qInv_times_q_mod_p =
-                    bigint::elem_mul(&qInv, q_mod_p, &p.modulus)?;
-                if !qInv_times_q_mod_p.is_one() {
-                    return Err(error::Unspecified);
-                }
+                let qInv = if pq_swapped {
+                    // Calculate a new qInv. Steps 7.c-7.f can be skipped because
+                    // elem_inverse() already verifies qInv's invertibility.
+                    bigint::elem_inverse(
```

Comment:
IIRC, we cannot use `elem_inverse` here because `elem_inverse()` is not constant time. Instead we need to use Fermat's Little Theorem: calculate qInv = `q**(p - 2)`. (We probably should rename `elem_inverse()` to `elem_inverse_vartime()` or something like that.)

Reply:
It seems reasonable to skip the check that q * qInv = 1 (mod p). However I think it's also reasonable to do the check anyway. In particular when we switch to Fermat's Little Theorem it's probably good to do the check since it wouldn't be (as) redundant.